### ISSUE TITLE
RFC: Improve CI and add more status items

### ISSUE THREAD

**Issue by nashif** [zephyrproject-rtos/zephyr#11501]
## Current Status

Currently all of our CI is run in one job and all results are reported via one status from shippable. We often have to wait more than 30 minutes to get a report that only complains about bad commit messages and such. Thos reports are hidden in  the over all report and not immediately visible to new comers and first time contributors.

## Proposal

- Move the compliance tests to a separate job that runs in parallel and reports to Github individually per test case (compliance and documentation tests).

- Report the output of the test to GH directly as a comment, this is visible to the authors directly, no need to go search for the report on shippable

- Each test has a status with a link to the guidelines

![screenshot 2018-11-19 at 10 01 42](https://user-images.githubusercontent.com/180017/48715101-3a6a9580-ebe2-11e8-9578-cc5997252220.png)

- Comments coming from CI will be updated for each run, i.e. we will not create new comments but edit one main comment only with CI results.

![screenshot 2018-11-19 at 10 07 21](https://user-images.githubusercontent.com/180017/48715405-f035e400-ebe2-11e8-96e2-40a7f83a8b24.png)

- All the compliance tests are run within ~10 minutes and reported to the PR as a comment before the actual sanitycheck run is done, giving developer immediate feedback.

## Things to consider

- Currently the recheck command in comments will only work with sanitycheck and full CI run, not the compliance statuses

- Documentation build is a bit slow, so the compliance job will take approx 10 min, this needs to be optimized in the future

- Documentation build is moved out of the main job into the new one, leaving the main CI job only for running sanitycheck.

**Comment by galak** [comment]
Can we add recheck-{checkpatch, doc, gitlint, etc.., all}?

**Comment by ceolin** [comment]
Can we add soft checks ? We have some coccinelle scripts that would be
great to be checked. Some of them are testing MISRA-C rules and it'd
nice to have a report. The problem is that they run per file/directory
so we can have issues not related with the patch, also it can have
false-positives. Though, this still can be used to help on review.

**Comment by galak** [comment]
> Can we add soft checks ? We have some coccinelle scripts that would be
> great to be checked. Some of them are testing MISRA-C rules and it'd
> nice to have a report. The problem is that they run per file/directory
> so we can have issues not related with the patch, also it can have
> false-positives. Though, this still can be used to help on review.

I wondering about using something like https://lgtm.com/ for such checks.

**Comment by nashif** [comment]
> Can we add recheck-{checkpatch, doc, gitlint, etc.., all}?

i would prefer to have 1 single recheck that does everything and that will be done but not immediately, there is no value on doing them individually, it will take almost exactly the same time.

**Comment by nashif** [comment]
> Can we add soft checks ? We have some coccinelle scripts that would be
> great to be checked. Some of them are testing MISRA-C rules and it'd
> nice to have a report. The problem is that they run per file/directory
> so we can have issues not related with the patch, also it can have
> false-positives. Though, this still can be used to help on review.

yes, this can be done without reporting a status, this can probably just generate a report and post it somewhere for review. We can always run this on master and have a baseline and compare the results of the PR with master and report the diff for example.

**Comment by nashif** [comment]
done already, so closing.
